packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts

```diff
@@ -1,30 +1,303 @@
+import assert from "assert";
+import chalk from "chalk";
+import { Mutex } from "miniflare";
+import {
+	createPreviewSession,
+	createWorkerPreview,
+} from "../../dev/create-worker-preview";
+import {
+	createRemoteWorkerInit,
+	getWorkerAccountAndContext,
+	handleUserFriendlyError,
+} from "../../dev/remote";
+import { UserError } from "../../errors";
+import { logger } from "../../logger";
+import { getAccessToken } from "../../user/access";
 import { RuntimeController } from "./BaseController";
+import { castErrorCause } from "./events";
 import { notImplemented } from "./NotImplementedError";
+import {
+	convertBindingsToCfWorkerInitBindings,
+	MissingConfigError,
+	unwrapHook,
+} from "./utils";
+import type {
+	CfPreviewSession,
+	CfPreviewToken,
+} from "../../dev/create-worker-preview";
+import type { ParseError } from "../../parse";
 import type {
 	BundleCompleteEvent,
 	BundleStartEvent,
 	PreviewTokenExpiredEvent,
 	ReloadCompleteEvent,
 	ReloadStartEvent,
 } from "./events";
+import type { Trigger } from "./types";
 
 export class RemoteRuntimeController extends RuntimeController {
+	#abortController = new AbortController();
+
+	#currentBundleId = 0;
+	#mutex = new Mutex();
+
+	#session?: CfPreviewSession;
+
+	async #previewSession(
+		props: Parameters<typeof getWorkerAccountAndContext>[0]
+	): Promise<CfPreviewSession | undefined> {
+		try {
+			const { workerAccount, workerContext } =
+				await getWorkerAccountAndContext(props);
+
+			return await createPreviewSession(
+				workerAccount,
+				workerContext,
+				this.#abortController.signal
+			);
+		} catch (err: unknown) {
+			assert(err && typeof err === "object");
+			// instead of logging the raw API error to the user,
+			// give them friendly instructions
+			// for error 10063 (workers.dev subdomain required)
+			if ("code" in err && err.code === 10063) {
+				const errorMessage =
+					"Error: You need to register a workers.dev subdomain before running the dev command in remote mode";
+				const solutionMessage =
+					"You can either enable local mode by pressing l, or register a workers.dev subdomain here:";
+				const onboardingLink = `https://dash.cloudflare.com/${props.accountId}/workers/onboarding`;
+				logger.error(`${errorMessage}\n${solutionMessage}\n${onboardingLink}`);
+			} else if (
+				"cause" in err &&
+				(err.cause as { code: string; hostname: string })?.code === "ENOTFOUND"
+			) {
+				logger.error(
+					`Could not access \`${(err.cause as { code: string; hostname: string }).hostname}\`. Make sure the domain is set up to be proxied by Cloudflare.\nFor more details, refer to https://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route`
+				);
+			} else if (err instanceof UserError) {
+				logger.error(err.message);
```

Comment:
should we prefix this error as well, similarly to [here](https://github.com/cloudflare/workers-sdk/pull/5939/files#diff-f3311016ad4bd6233c4740b176f304287ceeb2d88e865bdfed4ba09d0eaa3087R83)?
```
logger.error("Error while creating remote dev session:", err);
```

Reply:
Potentially, but let's address this in a followup since the messaging is the same as the existing flow